WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,273 @@
+        ///
+        /// - parameter sizeLimit: Size limit in bytes. `100 Mb` by default.
+        public init(sizeLimit: Int = 1024 * 1024 * 100) {
+            self.sizeLimit = sizeLimit
+        }
+    }
+
+    /// Creates a cache instance with a given `name`. The cache creates a directory
+    /// with the given `name` in a `.cachesDirectory` in `.userDomainMask`.
+    public init(name: String, configuration: Configuration = .init()) {
+        let cachesURL = DiskCache.getCachesURL()
+        self.init(url: cachesURL.appendingPathComponent(name, isDirectory: true), configuration: configuration)
+    }
+
+    private static func getCachesURL() -> URL {
+        guard let cachesURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            DDLogError("Failed to instantiate cache. Caches directory not available")
+            return URL(fileURLWithPath: "/dev/null") // This should never happen
+        }
+        return cachesURL
+    }
+
+    /// Creates a cache instance with a given root URL.
+    public init(url: URL, configuration: Configuration = .init()) {
+        self.rootURL = url
+        self.configuration = configuration
+
+        do {
+            try FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            DDLogError("Failed to creates cache root directory at: \(url) with error: \(error)")
+        }
+        DispatchQueue.main.asyncAfter(deadline: .now() + configuration.initialSweepDelay) { [weak self] in
+            Task {
+                await self?.performAndScheduleNextSweep()
+            }
+        }
+    }
+
+    // MARK: Codable
+
+    public func getValue<T: Decodable>(
+        _ type: T.Type,
+        forKey key: String,
+        decoder: JSONDecoder = JSONDecoder()
+    ) async -> T? {
+        guard let data = getData(forKey: key) else { return nil }
+        return try? decoder.decode(type, from: data)
+    }
+
+    public func setValue<T: Encodable>(
+        _ value: T,
+        forKey key: String,
+        encoder: JSONEncoder = JSONEncoder()
+    ) async {
+        guard let data = try? encoder.encode(value) else { return }
+        setData(data, forKey: key)
+    }
+
+    public func removeValue(forKey key: String) {
+        removeData(forKey: key)
+    }
+
+    // MARK: Codable (Closures)
+
+    public nonisolated func getValue<T: Decodable>(
+        _ type: T.Type,
+        forKey key: String,
+        decoder: JSONDecoder = JSONDecoder(),
+        _ completion: @escaping (T?) -> Void
+    ) {
+        Task {
+            let value = await getValue(type, forKey: key, decoder: decoder)
+            completion(value)
+        }
+    }
+
+    public nonisolated func setValue<T: Encodable>(
+        _ value: T,
+        forKey key: String,
+        encoder: JSONEncoder = JSONEncoder()
+    ) {
+        Task {
+            await setValue(value, forKey: key, encoder: encoder)
+        }
+    }
```

Comment:
Even though this is not a Swift async function, the cache file is still saved asynchronously. That means there might be a race condition between this API and the `async getValue`(and `async getData`).

This test case fails on my Mac: the assertions fails about 10 out of those 100 iterations.

```swift
    func testSetValueAPI() async {
        struct Payload: Codable {
            let value: String
        }
        let callSetValue: (Encodable, String) -> Void = { payload, key in
            self.cache.setValue(payload, forKey: key)
        }
        for index in 0...100 {
            let random = String("abcdefghijklmnopqrstuvwxyz".randomSample(count: 100_000))
            let payload = Payload(value: random)
            let key = "key-\(index)"
            callSetValue(payload, key)
            let data = await cache.getValue(Payload.self, forKey: key)
            XCTAssertEqual(data?.value, random)
        }
    }
```

Reply:
This version of setValue no longer exists in the updated and simplified implementation.